### DISCUSSION TITLE
ARM Directories

### DISCUSSION THREAD

**Discussion by trleith** [automatic-ripping-machine/automatic-ripping-machine#1003] [Ideas]
This is rambly and tenative.

I'm a novice -- ARM is the only reason I ran Docker at first. I'm absolutely not good at any of it.

In a former life, I did some SysAdmin work, technical writing, and training. FWIW. I wasn't good at any of that either.

Is a suggestion like this welcome? If so, what do people think?

I know it is't technically correct, but I'm going to call the Docker Container/Image the Virtual Machine (VM) and the Docker host the Real Machine (RM)

The Docker Image has

```
 "Volumes": {
                "/etc/arm/config": {},
                "/home/arm/Music": {},
                "/home/arm/logs": {},
                "/home/arm/media": {}
            },
            "WorkingDir": "/home/arm"

```
The docker run command I was given has

 ```
--mount type=bind,source=/<path on RM>,target=/home/arm
 --mount type=bind,source=/<path on RM>/media,target=/home/arm/media
 --mount type=bind,source=/<path on RM>/media/completed,target=/home/arm/media/completed
 --mount type=bind,source=/<path on RM>/media/movies,target=/home/arm/media/movies
 --mount type=bind,source=/<path on RM>/media/tv,target=/home/arm/media/tv
 --mount type=bind,source=/<path on RM>/media/music,target=/home/arm/media/Music
 --mount type=bind,source=/<path on RM>/media/raw,target=/home/arm/media/raw
```

What I actually see on my machine when ARM is running is confusing; ARM creates directories on its own in places I don't expect. It makes a link between Music and music. Why am I defining media/completed (which gets movies created automagically underneath it) and also defining media/movies which isn't used? I don't see a good correspondence between the Docker Image Definition and the Docker Container Definition.

Here's what I think I'd like to accomplish out in the Real Machine

1. Separate the guts of ARM from the completed output
2. Preserve arm.yaml and abcde.conf in the RM even if the VM is deleted, to avoid the current backup/restore dance
3. Never ever have a link between directories out in the RM, especially the output space
4. Consolidate output under a single directory (get rid of Music in /home/arm)

ARM (I think) needs three things:

- A Data Space on the RM where the completed work is dropped for the user
- A VM Space on the RM for ARM to communicate other things to the user and for the user to provide scratch space
- A Linux environment in the VM the user "shouldn't" have to be aware of

I'm thinking of something like this as the view within the Docker Container

```
/home/arm [I *guess* /home/arm is to make a shell inside the running Docker VM work right]
   arm_vm (directory)
      config (directory)
        arm.yaml
        abcde.cong
        .MakeMKV
      db
        arm.db
      logs (directory)
        <various log files>
      tmp (directory)
        transcode (directory)
        raw (directory -- why do we need both raw and transcode?)
           .dvdcss
           <wav, mkv, whatever>
   arm_completed (directory)
      music (directory)
      movies (directory)
      tv (directory)
      unknown (directory)
```

Then we should be able to define only two paths in the Docker Container Definition. 
I _think_ this is deprecated Docker syntax. The project should provide the Best Practice as a template

```
  --mount type=bind,source=/<path on RM>,target=/home/arm 
  --mount type=bind,source=/<path on RM'>,target=/home/arm/arm_completed
```

When ARM starts up, it can build the rest of the tree it needs, and put its template config files out in RM-land pre-configured, not in the VM's /etc/arm/config. The template configurations should match the way ARM is actually installed, but provided ONLY if there's no arm.yaml or abcde.conf already existing -- don't overwrite my work. <path on RM> and <path on RM'> could be the same or not. The VM won't know.

This seems easier to explain to people -- arm_vm is the guts of ARM; arm_completed is where you find your processed data, there are no other directories and no surprises. You can map them to the same directory on the RM filesystem or not. In the VM it all appears under /home/arm.

- media goes away
- Music goes away
- config is moved out into the RM/User Space
- no links created to confuse people (or Samba?)
- Absolutely no directories distinguished only by case (music/Music)
- Nothing unnecessary gets built

Ideally, if you ls -l /home/arm you see at most two directories and a README.txt file

The path configurations in ARM Settings could move to "Advanced" or something. *THIS* *might* build links in the RM filesystem. It's Advanced.

**Comment by koeberlue** [comment]
I also found some of the directory logic kind of confusing. So for example arm always creates the empty `raw` and `transcode` folder on my NAS, where the completed movies reside,  although they are mounted explictly on the working machine.

Concerning your proposal, I do have some additions:

* I would like to explicitly mount the `tmp` folder as well, because it will need to have enough space. If I am not mistaken, docker would by default store those files in memory, which would increase the hardware requirements a lot.
* IMO it should be possible (but not mandatory) to mount the subdirectories of your proposed `arm_completed` folder to different places. People do have different requirements on where to store each kind of media for playback. So e.g. I want to store movies somewhere for Jellyfin to pick them up while I want music to be stored somewhere else for moOde to pickt it up.

**Reply by trleith** [reply]
Thank you for your comment.

Your use-cases would be covered by creating links in the RM File System, presumably with the ARM Settings tool. Links you create yourself and are warned about are tolerable (IMO). I don't think the system should expect naive users to deal with links; asking them to follow a simple Docker recipe is enough.

The other way around this I suppose is a complex Docker recipe. Well, a longer Docker recipe -- it really isn't any more complex. I don't know at this time if I specified a mapping to the RM for /home/arm/completed/movies in the Docker Container Definition what would happen -- ARM creates that directory itself. Maybe if it finds it already existing because Docker created it, it'll just use it. But I forget whether defining a directory in a Docker Container Definition actually creates the directory, or if that's a separate step in the shell. I may try it to see what happens.

If I understand correctly, if you wanted to use a RAM Disk (or virtual memory) for the transcoding buffer, you're going to need 2X your media capacity per simultaneous transcode -- for a Blu Ray upwards of 200GB (each). That's a lotsa RAM. A conundrum, for sure.
